spec/rollout_spec.rb

```diff
@@ -89,17 +89,54 @@
     end
   end
 
-  describe "activating a specific user" do
-    before do
-      @rollout.activate_user(:chat, double(id: 42))
+  ##
+  # CONTRACT: assess feature status for the user
+  #
+  # assumming client uses it like this rollout.active?('feature_name', user)
+  #
+  # accepts feature name as a string and a user (an object with id attribute)
+  #
+  # returns boolean: true or false
+  # * true when feature is active for the user
+  # * false when feature isn't active for the user
+  ##
+  shared_examples "User's Feature Checker" do
+    def rollout(feature_name, user)
+      raise 'Subclass Responsibility'
     end
 
-    it "is active for that user" do
-      expect(@rollout).to be_active(:chat, double(id: 42))
+    def rollout_except(feature_name, user)
+      raise 'Subclass Responsibility'
     end
 
-    it "remains inactive for other users" do
-      expect(@rollout).not_to be_active(:chat, double(id: 24))
+    context 'when feature is rolled out for the user' do
+      subject { rollout('feature_name', double(id: 42)) }
+      it { is_expected.to be_active('feature_name', double(id: 42)) }
+    end
+
+    context 'when feature is NOT rolled out for the user' do
+      subject { rollout_except('feature_name', double(id: 42)) }
+      it { is_expected.not_to be_active('feature_name', double(id: 42)) }
+    end
+  end
+
+  describe "feature as a String and user as an anything with ID attribute" do
+    it_behaves_like "User's Feature Checker" do
+      def rollout(feature_name, user)
+        Rollout.new(Redis.new).tap do |rollout|
+          rollout.activate_user("not_the_#{feature_name}", user)
+          rollout.activate_user(feature_name, user)
+          rollout.activate_user(feature_name, double(id: user.id + 100))
```

Comment:
Should these three cases be translated into separate test cases? The first and third activation really check that there is no interference with other features/other users. This seem like first-class properties of the contract to me.

Reply:
Good question! I decided to give it a try, but it raises some more questions :)

For example, how do we express the case when feature and user both activated, but not for each other?

```ruby
context 'when feature and user activated, but not for each other' do
  subject { Rollout.new(Redis.new) }
  let(:feature_name) { 'feature_name' }
  let(:user) { double(id: 42) }

  before do
    subject.activate_user("not_the_#{feature_name}", user)
    subject.activate_user(feature_name, double(id: user.id + 100))
  end

  it { is_expected.not_to be_active(feature_name, user) }
end
```

Or when the feature was activated for the user, but than another feature was activated for the same user?

```ruby
context 'when another feature was also activated for the user' do
  subject { Rollout.new(Redis.new) }
  let(:feature_name) { 'feature_name' }
  let(:user) { double(id: 42) }

  before do
    subject.activate_user(feature_name, user)
    subject.activate_user("not_the_#{feature_name}", user)
  end

  it { is_expected.to be_active(feature_name, user) }
end
```

And so on...

It looks like we still need to combine cases in order to achieve the same level of confidence/completeness as we have with original factory method. I imagine if we continue combining those cases we eventually will uncover scenario which combines them all:

```ruby
context 'active pair among the other activations' do
  subject { Rollout.new(Redis.new) }
  let(:feature_name) { 'feature_name' }
  let(:user) { double(id: 42) }

  before do
    subject.activate_user("not_the_#{feature_name}", user)
    subject.activate_user(feature_name, user)
    subject.activate_user(feature_name, double(id: user.id + 100))
  end

  it { is_expected.to be_active(feature_name, user) }
end
```

I might be wrong with terminology, but this last scenario looks like a "story test" for multi user, multi feature implementation of rollover. And those individual test cases are probably how we would test drive implementation, i.e. triangulating different cases until we get to the "final" scenario.

And that's just one supplier, alternative suppliers might have different implementations, for example, single activation supplier where each new activation overrides previous one would require different test cases.

Does Client need to know about those details?

Assuming that the Client uses rollout like so `rollout.active?('feature_name', user)`, then the minimal details it needs to know is that for any given pair of User and FeatureName supplier will always return a boolean either true or false, where "true" represent "user has access to this feature" and false represents "user don't have access to this feature". And it looks like that is the only details we can put into Contract Test at the moment :)

Let me know your thoughts.

Thank you!

Reply:
First, let's agree that there's infinitely many possible sequences of `subject.active_user(a, b)` calls and that we cannot ever cover them all with tests. Therefore, our goal is a sample of sequences that gives us high confidence every concrete implementation that passes all tests fulfills the contract.

Second, we have to agree what our contract encompasses. (1) It's certainly that if "a feature" is activated for a user, it's active for this user. (2) It's probably also that if "a feature" is not activated for a user, it's not active for this user.

Note that case (2) does not require that "another feature" is activated. Strictly speaking we need no `activate_user` calls in the setup for case (2). With such a call, we get to case (3) that "a feature" is not active for a user if "another feature" is activated for this user. I would say this is also an important aspect of the contract. Let's call it "non-interference of features". There's infinitely many possible scenarios where interference could occur, so we cannot test them all. Since we can express this aspect with two features and I have no indication that there's anything special happening with more than two features, I would write the test with two features. (Another way to go might be property-based tests, but let's not go there just yet ;))

And then there's case (4) that "a feature" doesn't become active for a user if it is activated for a different user. I would say this is also an important aspect of the contract. Let's call it "non-interference of users". For the same reasons as above, I would write a test with just two users.

The last remaining case would be (5) that "a feature" doesn't become active for a user if "another feature" is activated for a different user. Though we certainly want this, I would probably not add a contract tests, because it seems unlikely that an implementation would break this, but not case (4).

So this leads me to four test cases in my contract test that cover the equivalence classes of scenario I see. The beauty about the particular tests I chose is that each fails _only_ for one particular reason: (1) if activation for a user does not work, (2) if non-activation for a user does not work, (3) if features interfere for a user, (4) if features interfere between users. If we combine any of the four test cases into a single test, then this composed test still does not test anything more than separate ones, but if it fails, we have multiple possible reasons to assess.

Does this make sense to you?

Reply:
> "a feature" is not active for a user if "another feature" is activated for this user. I would say this is also an important aspect of the contract. Let's call it "non-interference of features".

May I ask you to elaborate on how you decide what is an important aspect of the contract? :)
Which criteria could I use to decide which details is important for the "contract tests" and which are just specific implementation details of particular supplier, i.e. "specific supplier tests"?

Thank you in advance!

Reply:
> In the end, it depends on the business requirements. My judgement comes from assumptions I made about them

Interesting! I treated those Contract Tests as a part of design technique to express relationships between components, I'd probably have separate tests to express assumptions about particular supplier choice, for example "Learning Test" when using 3rd party library.

From this perspective it seems to me that adding those details into Contract Test would make design more rigid with no extra value in return. For example, we might have a Client which uses Supplier like so:
 
```ruby
if rollout.active?(new_feature.name, current_user)
  # do one thing
else
  # do another thing
end
```

practically it can work with many kind of suppliers as long as they shares common interface, for example (seriously, those are the different suppliers):

* multiple users can have access to multiple features simultaneously (current)
* multiple features, but each can only be accessed by single user at a time, the last activation wins
* same as previous, but first activation wins and can't be overridden
* how about limited time offer activations?
* or random % activations?
* and so on

Those suppliers will need to account for the different edge cases with their own tests (or learning tests), but if we include extra details about any particular Supplier into the Contract Test it will looks as if the Client declines to work with other kind of suppliers even though he really can, i.e. syntax and semantic are the same. Why? :)

Does that make more sense now not to include those details or what problems do you see with Contract Tests if we don't include proposed details? :)

Let me know your thoughts.

Thank you!

Reply:
> I think we have a different understanding of what activate_user() means

possibly :)

I'm trying to describe contract of `active?('a', 'b')` without concerning with the way particular implementation strategy activates users/features. The abstract factory methods (`rollout` and `rollout_except`) only defined in the contract tests (`shared_examples "User's Feature Checker"`), but implemented by the tests of particular strategy, i.e. different suppliers tests of different strategies can have a different implementations of the factory methods and can also implement more granular tests cases as you proposed.

> I'm not sure what you mean by contract test express relationships between components. In what sense does a tests express a relationship?

In a sense that Collaboration Tests for the Client gives the clue about Contract Tests for Supplier, as per my understanding of exercise:

> If you already have Collaboration Tests for the Client, then look at its stubs and expectations for clues about which Supplier details it needs to know. Remember that these stubs and expectations are assumptions about how the Supplier behaves.
> https://online-training.jbrains.ca/courses/8788/lectures/9607445

Given the Client with the following logic:

```ruby
if @rollout.active?(feature_name, user)
  # unlock the feature
else
  # display error message
end
```

Collaboration Tests would probably stub "rollout" signalling "true" in one case and "false" in another. Which gives me a clue that Client needs to know that for any given pair of Feature Name and User, `@rollout.active?(feature_name, user) will return either "true" or "false", where "true" mean feature is active, and "false" means opposite.

Why would we need to constrain Supplier beyond those details? Shouldn't business requirements for a specific Supplier be concern of that Supplier implementation?

I suppose there might be cases where we need to add more details into Contract Test which are not in Collaboration Test, for example, when Client can suddenly send NULL, and it is not in our control to prevent this, so, we would like Supplier to deal with it, which seems ok to me, but that's looks different from the business cases you proposed, I'm not saying that those are wrong or bad ideas, just trying to better understand weather I'm missing something  about Contract Tests :)

Reply:
> Why would we need to constrain Supplier beyond those details? Shouldn't business requirements for a specific Supplier be concern of that Supplier implementation?

To the second question, "Yes", if I understood the question. I would expect the implementation of `active?()` to hide the details of _how to decide_ whether feature `f` is active for user `u`. Maybe they decide by some role-based authorization policy or by looking at the calendar because `u` should only do `f` on Thursdays.

The overall application cares about this detail (it chooses the OnlyOnThursdaysRollout class) and the Rollout implementation tests care (they check what happens on every day of the week and the boundaries around midnight and daylight-saving time), but no other part of the system cares.

I would expect a Contract Test that shows that:

* `active?()` can return true or false and nothing else
* `active?()` can only raise a generic error type (this seems implicit in the contract of every function that doesn't otherwise specify errors that it might raise)

The signature of `active?()` carries the information that it needs at least a feature and a user to decide.

That sounds like everything to me!

Reply:
> I'm trying to describe contract of `active?('a', 'b')` without concerning with the way particular implementation strategy activates users/features. The abstract factory methods (`rollout` and `rollout_except`) only defined in the contract tests (`shared_examples "User's Feature Checker"`), but implemented by the tests of particular strategy, i.e. different suppliers tests of different strategies can have a different implementations of the factory methods and can also implement more granular tests cases as you proposed.

Ok, there we unveiled the difference in our understanding. I assumed that activating features is also part of the contract and you want to ensure a certain relation between activation and `active?()`. If it isn't, I agree with your set of contract tests.

Thanks for this lesson in challenging our implicit assumption and getting to a common understanding.